### ISSUE TITLE
Missing syntax highlighting for several languages

### ISSUE THREAD

**Issue by Steffan153** [exercism/exercism#5801]
[ ] Prolog
[ ] PHP
[ ] CFML

Will update this if I find more.

Something of help for Prolog? https://github.com/SWI-Prolog/swish/blob/master/web/js/codemirror/mode/prolog/prolog.js

**Comment by joshgoebel** [comment]
@iHiD What is our baseline browser requirement for Exercism v3?  Textmate grammars might be a way to fill in missing gaps in our editing support (if it proved to work well), but I saw this:

> This package will only work in browsers with WebAssembly support.

**Comment by iHiD** [comment]
Erm, modern. Like 2021 releases of things? 

Weird that it needs it though! But sort of cool!

**Comment by joshgoebel** [comment]
>  Like 2021 releases of things?

Well then I'd say we'd be safe assuming web assembly :)  So possibly worth looking into.

<img width="1382" alt="Screen Shot 2021-09-07 at 10 57 07 AM" src="https://user-images.githubusercontent.com/6473/132366618-afac56f7-4433-4137-a75d-a72467b2c731.png">

**Comment by iHiD** [comment]
Nice.

**Comment by Steffan153** [comment]
Yeah, imo it's honestly not worth supporting IE anymore. Microsoft is fully ending support for it next year, Windows 11 doesn't even have IE, etc. Other than IE it's pretty awesome support.

**Comment by SleeplessByte** [comment]
"worth it" is really more a question if we have a significant user base still using those browsers (and I don't think we do), so I agree that we don't need to support it.

Also <https://death-to-ie11.com/>

**Comment by joshgoebel** [comment]
I find IE11 also kind of very difficult to support for a -modern- site these days without a dedicated IE11 champion and tester. So if we don't have such a person it's kind of also a moot point IMHO.

**Comment by Steffan153** [comment]
Did even Exercism 2 support IE11? :thinking:

**Comment by joshgoebel** [comment]
I think I forgot how old Web Assembly is now when I asked. :)  But I've also never heard us official say what we support browser wise so I had two reasons for asking.

**Comment by kntsoriano** [comment]
- I think for Prolog, we can’t use the current tokenizer because the tokens it produces doesn’t match the CodeMirror 6 highlighter.

- PHP is fairly easy to implement although I don’t know the language enough to do so. It is based on the clike tokenizer. The clike tokenizer has some parameters to configure it, but I don’t know what to put in there.

- I don’t know how to make CFML work with CodeMirror.

**Comment by Steffan153** [comment]
Afa the PHP, its built in to CodeMirror, so you shouldn't need to know any PHP to use it?

For CFML, I'll try it myself sometime and tell yall if it works.

For Prolog, I thought https://github.com/SWI-Prolog/swish/blob/master/web/js/codemirror/mode/prolog/prolog.js would work fine with the legacy highlighter.

**Comment by kntsoriano** [comment]
The tokens don't really match for the highlighter. It produces tokens such as `fullstop` and `qatom` which we don't have for the default codemirror syntax highlighter.

PHP isn't built in into CodeMirror 6, and there is no legacy support for it yet. I know that CodeMirror 5 uses the `clike` mode for it, but using that mode entails using some defaults that I don't know since I haven't used PHP.

**Comment by Steffan153** [comment]
PHP is builtin to CodeMirror 6: https://github.com/codemirror/lang-php

**Comment by kntsoriano** [comment]
@Steffan153 Thank you so much! I will add PHP now :)

**Comment by iHiD** [comment]
That's encouraging that they've added PHP support to CM6!

**Comment by Steffan153** [comment]
Hey, so I got codemirror textmate working, but I also found that the CF thing I found is actually for CF + HTML, and then for the <cfscript>, it just uses JS syntax highlighting because it's so close. Also, some other platforms like codewars just use JS as syntax highlight for CFML, so I think that should work just fine. I'll work on the textmate prolog though.

**Comment by Steffan153** [comment]
Here ya go! https://github.com/Steffan153/cm-textmate-prolog

I used Parcel so it would be nice and quick. The textmate files are in the dist. I just used https://marketplace.visualstudio.com/items?itemName=Togusa09.tmlanguage to convert the .tmLanguage files to JSON. Works pretty nice!

**Comment by Steffan153** [comment]
Added the rest of missing languages to the list and solutions.

**Comment by joshgoebel** [comment]
To be clear are we only talking about EDITING here?  If we're talking about editing and display then we need to be clear about CodeMirror (editor) vs Highlight.js (static display) support... We definitely have Highlight.js grammars for a lot of those (not sure about all) but I just wanted to make sure we're having the full discussion here.

**Comment by Steffan153** [comment]
Yep, only the editing.

**Comment by kntsoriano** [comment]
@Steffan153 Really great work on this. Thank you so much!

You deserve the Exercism rep by submitting a PR as you've done almost all the work here. I'd be glad to walk you through the ff:

1. Setting up the development environment.
2. Adding syntax highlighting to our code editor.

We can start with PHP as it should be the easiest.

**Comment by Steffan153** [comment]
Sure, I would have earlier but it said it wasn't open for PRs.

I'm horrible at getting Docker to work, I usually have to do 3 hours of StackOverflowing :joy:

Probably the main thing I need help with is the setting up of the dev env, because usually 95% is just errors. :joy:
I can start with it though and see how far I'll get.

**Comment by iHiD** [comment]
We can also add @Steffan153 to be co-author of the actual PR if that's easier :)

**Comment by Steffan153** [comment]
Now that I see how easy the dev env is to setup, I should be able to get it rolling.

**Comment by Steffan153** [comment]
Hey, so the dev env is almost working. I'm running this in the latest Ubuntu WSL2. The server isn't working though. It just says "local.exercism.io didn't send any data." Any ideas?

**Comment by Steffan153** [comment]
![image](https://user-images.githubusercontent.com/40404519/132962677-e189953c-a82f-42c6-a3d8-5481f3da159d.png)
I restarted it and now it's a diffferent error...?

**Comment by Steffan153** [comment]
It started workng from one day to the next. Smh :joy:

**Comment by Steffan153** [comment]
Next problem: Confirmation email wont come.

**Comment by Steffan153** [comment]
The only tracks are C#, Ruby, and Prolog. How do I add the others?

**Comment by kntsoriano** [comment]
1. Inside`db/seeds.rb:75`, edit the array to include the track slugs you want to add.

> This example adds `php` and `prolog`:

```ruby
db/seeds.rb:75
next unless %w[ruby csharp prolog php prolog].include?(track_slug)
```

2. After editing the seeds run `db:migrate:reset db:seed` in your terminal.

Do note, though, that the `iHiD` user has the password reset to the default. The default password is `password`.

**Comment by Steffan153** [comment]
So now, how do I access the files inside the container so I can use VScode, git, etc? I've just used the ./bin/shell before but it's inside Docker. I tried some code online for making a volume and it just spat errors at me. :joy:

**Comment by kntsoriano** [comment]
Hi @Steffan153, sorry about the long process of getting set up! Hopefully this should fix your issue:

Inside `development-environment/stack.yml`, you can edit the `website` block to be this:

```yml
# development-environment/stack.yml
  website:
    source: true
    environment:
      WEBPACKER_DEV_SERVER_PUBLIC: "local.exercism.io:9999"
      CABLE_URL: "ws://local.exercism.io:9998/cable"
    ports:
      - 3020:3020 # rails, 3020 internal
      - 9999:3035 # webpack, 3035 internal
      - 9998:3334 # cable, 3334 internal
```

The only thing I've changed above is that I've set `source` to true. This configuration means that your local version, located in `../website` (relative to `development-environment`) will be used.

You can now change code inside `../website` and your changes should now appear.

**Comment by Steffan153** [comment]
Didn't work. :/ Ykw, can we just do the co-author? I'm honestly getting a bit tired of errors.

**Comment by iHiD** [comment]
Yeah - getting this setup isn't trivial. @kntsoriano can you get the PR in and add @Steffan153 as add a note that I should add @Steffan153 as the co-author pls?

**Comment by kntsoriano** [comment]
Unfortunately, PHP highlighting breaks the syntax highlighter for Elixir as they’re both on incompatible versions. I figure that we just skip PHP for now as Elixir is a more popular track? Is this okay?

**Comment by Steffan153** [comment]
PHP is the one I was really wanting. :)

**Comment by kntsoriano** [comment]
@Steffan153 Yes. Some of our languages depend on the 0.18 version, while php is only compatible with 0.19. I've tried upgrading some of our dependencies but they break.

I will try this again today as I might've done something wrong when I was resolving some of our dependencies.

**Comment by Steffan153** [comment]
Vimscript found! https://github.com/skammer/textmate-viml/blob/master/VimL.tmbundle/Syntaxes/VimL.tmLanguage

(Textmate)

**Comment by kntsoriano** [comment]
@Steffan153 Thanks! I'm having difficulty integrating Textmate with CodeMirror 6. I'm unsure if it's supported as of the moment.

I've looked into https://github.com/Steffan153/cm-textmate-prolog and it seems to be on CodeMirror 5. Please let me know if I've missed anything here.

**Comment by joshgoebel** [comment]
If you all exhaust the options here for existing stuff and someone is ready to put in some work on any of these languages from scratch I'm open to being a resource.  I think we should perhaps bootstrap a few *simple* CM6 grammars and focus purely on KSC (keywords, strings, comments) and editor behavior (proper tabbing, etc)... Then we'd have a template others could use to at least get basic keyword highlighting working.

I can help a bit to get someone motivated started, but I don't have time to spearhead this.

**Comment by neenjaw** [comment]
@jiegillet enjoys parsing exercises, he might be interested language grammars. Not sure if he is aware of this current need.
